### PR TITLE
[SCB-12] Support Custom Metrics

### PR DESCRIPTION
**Pull request by zhengyangyong**
Signed-off-by: zhengyangyong <yangyong.zheng@huawei.com>

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean install` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
Add custom metrics interface for user define his/her own business metrics,also a sample in "custom-business-metrics"
```java
@Component
public class ShopDemoService {

  private final CounterService counterService;

  private final GaugeService gaugeService;

  private final WindowCounterService windowCounterService;

  //autowire metrics service
  @Autowired
  public ShopDemoService(CounterService counterService, GaugeService gaugeService,
      WindowCounterService windowCounterService) {
    this.counterService = counterService;
    this.gaugeService = gaugeService;
    this.windowCounterService = windowCounterService;

    login(null,null);
  }

  public void login(String name, String password) {
    counterService.increment("Active User");
  }

  public void logout(String session) {
    counterService.decrement("Active User");
  }

  public void order(double amount) throws InterruptedException {
    long start = System.currentTimeMillis();
    //sim  do order process
    Thread.sleep(100);

    //sim record order process time
    windowCounterService.record("Order Latency", System.currentTimeMillis() - start);

    //plus one
    windowCounterService.record("Order Count", 1);

    //only support long,please do unit convert ,$99.00 dollar -> $9900 cent, $59.99 dollar -> $5999 cent
    windowCounterService.record("Order Amount", (long) round(amount * 100, 0));
  }

  public void discount(double value) {
    //make a discount to Levis Jeans

    //record current Levis Jeans Discount
    gaugeService.update("Levis Jeans Discount", value);
  }

  private double round(double value, int places) {
    if (!Double.isNaN(value)) {
      BigDecimal decimal = new BigDecimal(value);
      return decimal.setScale(places, RoundingMode.HALF_UP).doubleValue();
    }
    return 0;
  }

  /*  Output of RegistryMetric.customMetrics :
   *  Active User
   *  Order Latency (total,count,tps,rate,average,max,min)
   *  Order Count (total,count,tps,rate,average,max,min)
   *  Order Amount (total,count,tps,rate,average,max,min)
   *  Levis Jeans Discount
   */
}
```

### CODE REVIEW

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/custom/WindowCounterService.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.servicecomb.metrics.core.custom;
+
+/**
+ GaugeService is complex service for manage Window Time-related Step Counter,
```
[the remark this answers is not in the document]
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/custom/WindowCounterService.java]: Done

File: metrics/metrics-core/src/test/java/org/apache/servicecomb/metrics/core/TestHealthCheckerManager.java [hunk 2]
```diff
@@ -84,5 +97,8 @@ public void testRegistry() throws IOException {
     Assert.assertTrue(data.getInstanceId().equals("001"));
```
[the remark this answers is not in the document]
Comment (zhengyangyong) [reply metrics/metrics-core/src/test/java/org/apache/servicecomb/metrics/core/TestHealthCheckerManager.java]: Done

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/custom/GaugeService.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.servicecomb.metrics.core.custom;
+
+/**
+ GaugeService is simple service for manage basic guage,Window Time-unrelated,always get latest value
```
Comment (liubao68) [inline metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/custom/GaugeService.java]: Gauge， not guage

Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/custom/GaugeService.java]: Done

**Comment by zhengyangyong** [comment]
Rebase and resolve conflicts

**Pull request merged**